README.md
# Helix Micropub Endpoint

> A micropub publishing endpoint for Project Helix

## Status
[![codecov](https://img.shields.io/codecov/c/github/trieloff/helix-micropub-endpoint.svg)](https://codecov.io/gh/trieloff/helix-micropub-endpoint)
[![CircleCI](https://img.shields.io/circleci/project/github/trieloff/helix-micropub-endpoint.svg)](https://circleci.com/gh/trieloff/helix-micropub-endpoint)
[![GitHub license](https://img.shields.io/github/license/trieloff/helix-micropub-endpoint.svg)](https://github.com/trieloff/helix-micropub-endpoint/blob/master/LICENSE.txt)
[![GitHub issues](https://img.shields.io/github/issues/trieloff/helix-micropub-endpoint.svg)](https://github.com/trieloff/helix-micropub-endpoint/issues)
[![LGTM Code Quality Grade: JavaScript](https://img.shields.io/lgtm/grade/javascript/g/trieloff/helix-micropub-endpoint.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/trieloff/helix-micropub-endpoint)
[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)

## Installation

None. This is a service.

## Demo

https://vimeo.com/424564648

## Purpose

[Micropub](https://indieweb.org/Micropub) is an API for creating posts on Websites. It is supported by a number of clients, most notably [IA Writer](https://ia.net/writer).

`helix-micropub-endpoint` is a Micropub endpoint for GitHub that works great with [Helix Pages](https://www.hlx.page). If you quickly need to post something from a desktop or mobile client without using the GitHub online editor, you might want to give it a try.

The base URL `https://adobeioruntime.net/api/v1/web/trieloff/helix-micropub/publish@v1/<owner>/<repo>/<ref>` includes your GitHub username, repository and the branch where you want posts to show up. Typically that would be `master`.


## Usage

Try the service here:

```bash
curl https://adobeioruntime.net/api/v1/web/trieloff/helix-micropub/publish@v1/<owner>/<repo>/<ref>
```

### Posting with IA Writer

1. Open IA Writer
2. Go to Settings and Accounts

![](./docs/add-micropub.png)

3. Click the plus button and select Micropub

![](./docs/enter-token-manually.png)

4. Select "Enter Token Manually"
5. Paste the URL from above.
6. Paste a GitHub developer token that has access to the `repo` and `user` scope (it will commit in your name)

![](./docs/post-markdown.png)

7. You are ready to post.

## Development

### Deploying Helix MicroPub Endpoint

Deploying Helix MicroPub Endpoint requires the `wsk` command line client, authenticated to a namespace of your choice. For this project, I am using the `trieloff` namespace.

All commits to master that pass the testing will be deployed automatically. All commits to branches that will pass the testing will get commited as `/helix-micropub/publish@ci<num>` and tagged with the CI build number.


CHANGELOG.md
# [1.5.0](https://github.com/trieloff/helix-micropub-endpoint/compare/v1.4.4...v1.5.0) (2020-06-01)


### Features

* **markdown:** create better file names from first line of markdown ([ee4380d](https://github.com/trieloff/helix-micropub-endpoint/commit/ee4380d1c40a00ccd581455cc741da1e4c3cd7d7))

## [1.4.4](https://github.com/trieloff/helix-micropub-endpoint/compare/v1.4.3...v1.4.4) (2020-05-31)


### Bug Fixes

* **drafts:** drafts now work again for JSON posts ([c789551](https://github.com/trieloff/helix-micropub-endpoint/commit/c7895517b75cc8f59badf2c607bd6fb26b660a5e))

## [1.4.3](https://github.com/trieloff/helix-micropub-endpoint/compare/v1.4.2...v1.4.3) (2020-05-31)


### Bug Fixes

* **micropub:** make responses look more like micro.blog ([6e8456a](https://github.com/trieloff/helix-micropub-endpoint/commit/6e8456afb9d29d9153fda20c759a6a8a241133b3))

## [1.4.2](https://github.com/trieloff/helix-micropub-endpoint/compare/v1.4.1...v1.4.2) (2020-05-31)


### Bug Fixes

* **discovery:** remove .json from advertised URL ([28d08bf](https://github.com/trieloff/helix-micropub-endpoint/commit/28d08bfd108bb65500110d2ec45a0f9ec378213c))

## [1.4.1](https://github.com/trieloff/helix-micropub-endpoint/compare/v1.4.0...v1.4.1) (2020-05-31)


### Bug Fixes

* **index:** always return JSON for config requests ([8654abb](https://github.com/trieloff/helix-micropub-endpoint/commit/8654abba903171e3615711c1d401ddc1565b0e4e))

# [1.4.0](https://github.com/trieloff/helix-micropub-endpoint/compare/v1.3.0...v1.4.0) (2020-05-31)


### Features

* **micropub:** support posting drafts as JSON ([8ead0b3](https://github.com/trieloff/helix-micropub-endpoint/commit/8ead0b3f742fd6413f612b0e967eecbd583f6981))

# [1.3.0](https://github.com/trieloff/helix-micropub-endpoint/compare/v1.2.1...v1.3.0) (2020-05-31)


### Features

* **index:** show config JSON when requested ([c2d0083](https://github.com/trieloff/helix-micropub-endpoint/commit/c2d0083f3d7fda3ea34b17f8b16c3d00d29030af))

## [1.2.1](https://github.com/trieloff/helix-micropub-endpoint/compare/v1.2.0...v1.2.1) (2020-05-31)


### Bug Fixes

* **micropub:** remove space from discovery link ([0a0ddde](https://github.com/trieloff/helix-micropub-endpoint/commit/0a0dddebe706a58ca6561cedad2c3fe385e25647))

# [1.2.0](https://github.com/trieloff/helix-micropub-endpoint/compare/v1.1.1...v1.2.0) (2020-05-31)


### Features

* **index:** show documentation on GET requests ([d04c45a](https://github.com/trieloff/helix-micropub-endpoint/commit/d04c45a37caa661747188a64035d4c8e56dd9bf5))

## [1.1.1](https://github.com/trieloff/helix-micropub-endpoint/compare/v1.1.0...v1.1.1) (2020-05-31)


### Bug Fixes

* **index:** use correct status code ([c0d90df](https://github.com/trieloff/helix-micropub-endpoint/commit/c0d90dfd2280df7f8f34a70205debcebe26e6433))

# [1.1.0](https://github.com/trieloff/helix-micropub-endpoint/compare/v1.0.0...v1.1.0) (2020-05-31)


### Bug Fixes

* **drafts:** fix preview links ([bd82f42](https://github.com/trieloff/helix-micropub-endpoint/commit/bd82f4247fe5fabf88248dcb2d41f34ce183f0d8))


### Features

* **drafts:** link to preview in PR description ([95a860b](https://github.com/trieloff/helix-micropub-endpoint/commit/95a860bf86f38db4d254073007e694841ea3cd6f))

# 1.0.0 (2020-05-31)


### Bug Fixes

* **index:** fix posting without drafting ([3aae092](https://github.com/trieloff/helix-micropub-endpoint/commit/3aae09286021136dfbebf77ac522a163801603b0))


### Features

* **index:** basic micropub implementation ([76c4faf](https://github.com/trieloff/helix-micropub-endpoint/commit/76c4faf744f3808c88760b8af2bdfa96d72b1136))

## [1.8.30](https://github.com/adobe/helix-service/compare/v1.8.29...v1.8.30) (2020-05-26)


### Bug Fixes

* **deps:** update dependency @adobe/helix-epsagon to v1.3.2 ([#197](https://github.com/adobe/helix-service/issues/197)) ([1ba78e4](https://github.com/adobe/helix-service/commit/1ba78e4f0f0e0b09c7eb63f66254ba65b27e890c))

## [1.8.29](https://github.com/adobe/helix-service/compare/v1.8.28...v1.8.29) (2020-05-19)


### Bug Fixes

* **deps:** update dependency yaml to v1.10.0 ([#195](https://github.com/adobe/helix-service/issues/195)) ([992afbb](https://github.com/adobe/helix-service/commit/992afbbd0e0ad143acc50e4b6c7cc046a7b0ca65))

## [1.8.28](https://github.com/adobe/helix-service/compare/v1.8.27...v1.8.28) (2020-05-18)


### Bug Fixes

* **deps:** update dependency @adobe/helix-epsagon to v1.3.1 ([34e5191](https://github.com/adobe/helix-service/commit/34e51911e9a4e283a0da9e7c2b851b657cd174ff))

## [1.8.27](https://github.com/adobe/helix-service/compare/v1.8.26...v1.8.27) (2020-05-18)


### Bug Fixes

* **deps:** update dependency @adobe/helix-epsagon to v1.3.0 ([#193](https://github.com/adobe/helix-service/issues/193)) ([4701c82](https://github.com/adobe/helix-service/commit/4701c82546c7ffb314b667e107c7a8c0a1606c0f))

## [1.8.26](https://github.com/adobe/helix-service/compare/v1.8.25...v1.8.26) (2020-05-12)


### Bug Fixes

* **deps:** update dependency @adobe/helix-epsagon to v1.2.2 ([#190](https://github.com/adobe/helix-service/issues/190)) ([db1db11](https://github.com/adobe/helix-service/commit/db1db11b883caf62416182bcc9c4c05624fd9c3a))

## [1.8.25](https://github.com/adobe/helix-service/compare/v1.8.24...v1.8.25) (2020-05-07)


### Bug Fixes

* **deps:** update dependency @adobe/helix-epsagon to v1.2.1 ([#187](https://github.com/adobe/helix-service/issues/187)) ([2617660](https://github.com/adobe/helix-service/commit/2617660cc9d61475fb7799145e1509ab57ccf6a3))

## [1.8.24](https://github.com/adobe/helix-service/compare/v1.8.23...v1.8.24) (2020-04-17)


### Bug Fixes

* **cfg:** use helix-serives package ([#178](https://github.com/adobe/helix-service/issues/178)) ([25fa3bf](https://github.com/adobe/helix-service/commit/25fa3bfbb3d0f2b5b051372177b87c82a9e430b5))

## [1.8.23](https://github.com/adobe/helix-service/compare/v1.8.22...v1.8.23) (2020-04-14)


### Bug Fixes


test/index.test.js
/*
 * Copyright 2019 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

/* eslint-env mocha */

'use strict';

const assert = require('assert');
const index = require('../src/index.js').main;

describe('Index Tests', () => {
  it('Index function rejects unknown types', async () => {
    const result = await index({
      __ow_path: '/trieloff/helix-demo/master',
      __ow_headers: {
        authorization: `Bearer ${process.env.GITHUB_TOKEN}`,
      },
      'post-status': 'draft',
      content: `# Hello World

This is a test`,
    });

    assert.equal(result.statusCode, 400, result.body);
  }).timeout(20000);

  it('Index shows documentation on GET', async () => {
    const result = await index({
      __ow_path: '/trieloff/helix-demo/master',
      __ow_method: 'get',
      __ow_headers: {
      },
    });

    assert.equal(result.statusCode, 200, result.body);
    assert.equal(result.headers.link, '<https://adobeioruntime.net/api/v1/web/trieloff/helix-micropub/publish@v1/trieloff/helix-demo/master>; rel="micropub"');
  }).timeout(20000);

  it('Index shows config on GET ?q=config', async () => {
    const result = await index({
      __ow_path: '/trieloff/helix-demo/master',
      __ow_method: 'get',
      __ow_headers: {
      },
      q: 'config',
    });

    assert.equal(result.statusCode, 200, result.body);
    assert.ok(Array.isArray(result.body.destination), JSON.stringify(result.body));
  }).timeout(20000);

  it('Index shows config on GET ?q=config as JSON when JSON has been requested', async () => {
    const result = await index({
      __ow_path: '/trieloff/helix-demo/master',
      __ow_method: 'get',
      __ow_headers: {
        accept: 'application/json',
      },
      q: 'config',
    });

    assert.equal(typeof result, 'object', result);
    assert.ok(Array.isArray(result.destination), JSON.stringify(result));
  }).timeout(20000);

  it('Index function creates drafts', async () => {
    const result = await index({
      __ow_path: '/trieloff/helix-demo/master',
      __ow_headers: {
        authorization: `Bearer ${process.env.GITHUB_TOKEN}`,
      },
      'post-status': 'draft',
      h: 'entry',
      content: `# Hello World

This is a test`,
    });

    assert.equal(result.statusCode, 202, result.body);
  }).timeout(20000);

  it('Index function creates drafts from JSON', async () => {
    // captured from Proxie
    const body = {
      type: ['h-entry'],
      properties: {
        name: ['Testy test'],
        content: ['Testy test.'],
        'post-status': ['draft'],
      },
    };

    const result = await index({
      __ow_path: '/trieloff/helix-demo/master',
      __ow_headers: {
        authorization: `Bearer ${process.env.GITHUB_TOKEN}`,
      },
      ...body,
    });

    assert.equal(result.statusCode, 202, JSON.stringify(result.body));
    assert.ok(result.body.url, JSON.stringify(result.body));
    assert.ok(result.body.preview, JSON.stringify(result.body));
  }).timeout(20000);

  it('Index function creates posts', async () => {
    const result = await index({
      __ow_path: '/trieloff/helix-demo/master',
      __ow_headers: {
        authorization: `Bearer ${process.env.GITHUB_TOKEN}`,
      },
      h: 'entry',
      content: `# Hello World

This is a test`,
    });

    assert.equal(result.statusCode, 202, result.body);
  }).timeout(20000);
});


test/post-deploy.js

/*
 * Copyright 2020 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

/* eslint-env mocha */
/* eslint-disable no-unused-expressions */

const chai = require('chai');
const chaiHttp = require('chai-http');
const packjson = require('../package.json');

chai.use(chaiHttp);
const { expect } = chai;

function getbaseurl() {
  const namespace = 'trieloff';
  const package = 'helix-micropub';
  const name = 'publish';
  let version = `${packjson.version}`;
  if (process.env.CI && process.env.CIRCLE_BUILD_NUM && process.env.CIRCLE_BRANCH !== 'master') {
    version = `ci${process.env.CIRCLE_BUILD_NUM}`;
  }
  return `api/v1/web/${namespace}/${package}/${name}@${version}`;
}

describe('Post-Deploy Tests', () => {
  it('Publish a Blog Post (draft)', async () => {
    console.log(`Trying https://adobeioruntime.net/${getbaseurl()}/trieloff/helix-demo/master`);

    await chai
      .request('https://adobeioruntime.net/')
      .post(`${getbaseurl()}/trieloff/helix-demo/master`)
      .type('form')
      .set('Authorization', `Bearer ${process.env.GITHUB_TOKEN}`)
      .send({
        'post-status': 'draft',
        h: 'entry',
        content: `# My awesome blog post

Hello world from CircleCI. This is build # ${process.env.CIRCLE_BUILD_NUM}
`,
      })
      .then((response) => {
        expect(response).to.have.status(202);
      })
      .catch((e) => {
        throw e;
      });
  }).timeout(10000);
});


test/markdown.test.js
/*
 * Copyright 2019 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

/* eslint-env mocha */

const assert = require('assert');
const { plain, slug } = require('../src/index.js');

describe('Markdown Tests', () => {
  it('plain strips markdown', () => {
    assert.equal(plain('# This is *so* _cool_'), 'This is so cool\n');
  });

  it('slug creates a slug from Markdown', () => {
    const markdown = `

# Hello World

This is my test
`;
    assert.equal(slug(markdown, 'now'), 'hello-world-now');
  });
});


src/index.js
/*
 * Copyright 2019 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */
const { wrap } = require('@adobe/openwhisk-action-utils');
const { logger } = require('@adobe/openwhisk-action-logger');
const { wrap: statusCheck } = require('@adobe/helix-status');
const { epsagon } = require('@adobe/helix-epsagon');
const { Octokit } = require('@octokit/rest');
const remark = require('remark');
const strip = require('strip-markdown');
const GithubSlugger = require('github-slugger');
const AWS = require('aws-sdk');

const slugger = new GithubSlugger();

/* eslint-disable camelcase, no-param-reassign */

function plain(markdown) {
  return remark()
    .use(strip)
    .processSync(markdown)
    .contents;
}

function slug(markdown, now) {
  const first = plain(markdown).split('\n').find((line) => line.trim() !== '');
  return slugger.slug(`${first}-${now}`);
}

/**
 * Turn Microformats-like JSON into regular JSON
 * @param {*} params
 */
function cleanup(params) {
  return Object.entries(params).reduce((clean, [key, value]) => {
    if (key.startsWith('__')) {
      // pass through
      clean[key] = value;
    } else if (key === 'type' && Array.isArray(value) && value.length === 1 && value[0].startsWith('h-')) {
      clean.h = value[0].substring(2);
    } else if (Array.isArray(value) && value.length === 1) {
      const [myval] = value;
      clean[key] = myval;
    } else if (key === 'properties') {
      return Object.assign(clean, cleanup(value));
    } else {
      clean[key] = value;
    }
    return clean;
  }, {});
}

/**
 * This is the main function
 * @param {string} name name of the person to greet
 * @returns {object} a greeting
 */
async function main(params) {
  try {
    const now = Math.round((new Date()).getTime() / 1000);
    const year = new Date().getUTCFullYear();

    const {
      AWS_KEY,
      // eslint-disable-next-line camelcase
      __ow_path, h, content, __ow_headers: { authorization }, __ow_method, q,
    } = cleanup(params);

    const [owner, repo, base] = __ow_path.replace(/^\//, '').split('/');

    const auth = authorization.split(' ').pop();

    // assume everything is there
    const github = new Octokit({ auth, userAgent: 'helix-micropub-endpoint' });

    const user = (await github.users.getAuthenticated()).data;

    if (q === 'config') {
      const s3 = new AWS.S3({
        signatureVersion: 'v4'
      });

      const media = await s3.getSignedUrlPromise('putObject', {
        Bucket: 'helix-micropub-media-endpoint',
        Key: AWS_KEY,
      });

      return {
        statusCode: 200,
        headers: {
          'content-type': 'application/json',
        },
        destination: [{
          uid: `https://github.com/${owner}/${repo}`,
          name: `${owner}/${repo} on GitHub via Helix`,
        }],
        'post-types': [
          { type: 'note', name: 'Post' }],
        body: {
          'media-endpoint': media,
          destination: [{
            uid: `https://github.com/${owner}/${repo}`,
            name: `${owner}/${repo} on GitHub via Helix`,
          }],
          'post-types': [
            { type: 'note', name: 'Post' }],
        },
      };
    }
    if (__ow_method === 'get') {
      return {
        statusCode: 200,
        headers: {
          'content-type': 'text/html',
          link: `<https://adobeioruntime.net/api/v1/web/trieloff/helix-micropub/publish@v1/${owner}/${repo}/${base}>; rel="micropub"`,
        },
        body: `<html>
  <head>
    <title>Helix MicroPub Endpoint</title>
    <link rel="micropub" href="https://adobeioruntime.net/api/v1/web/trieloff/helix-micropub/publish@v1/${owner}/${repo}/${base}">
  </head>
  <body>
    This is a <a href="https://www.w3.org/TR/micropub/">Micropub</a> endpoint. It expects <code>POST</code>
    requests.
  </body>
</html>`,
      };
    }

    if (h !== 'entry') {
      return {
        statusCode: 400,
        body: `Unknown type ${h}`,
      };
    }
    const status = cleanup(params)['post-status'] || 'published'; // default is published

    const mybranch = await github.repos.getBranch({ owner, repo, branch: base });
    const { sha } = mybranch.data.commit;

    const head = status === 'draft' ? `newpost-${now}` : base;
    const ref = status === 'draft' ? `refs/heads/${head}` : `refs/heads/${base}`;

    if (status === 'draft') {
    // create the branch
      await github.git.createRef({
        owner,
        repo,
        ref,
        sha,
      });
    }

    const path = `${year}/${slug(content, now)}.md`;
    const html = `${year}/${slug(content, now)}.html`;

    // update the file
    await github.repos.createOrUpdateFile({
      owner,
      repo,
      branch: head,
      path,
      content: Buffer.from(content).toString('base64'),
      committer: {
        name: user.name,
        email: user.email,
      },
      author: {
        name: user.name,
        email: user.email,
      },
      message: 'New blog post by helix-micropub-endpoint',
    });

    if (status === 'draft') {
    // create a PR
      const result = await github.pulls.create({
        owner,
        repo,
        title: `New Blog Post ${new Date().toLocaleDateString()}`,
        body: `Please review my [new blog post](https://${head}--${repo}--${owner}.hlx.page/${html})
        
Just approve this PR to get the post published at https://${repo}-${owner}.hlx.page/${html}`,
        base,
        head,
      });

      return {
        statusCode: 202,
        headers: {
          // eslint-disable-next-line no-underscore-dangle
          Location: result.data._links.html.href,
        },
        body: {
          // eslint-disable-next-line no-underscore-dangle
          url: result.data._links.html.href,
          preview: `https://${head}--${repo}--${owner}.hlx.page/${html}`,
        },
      };
    }

    return {
      statusCode: 202,
      headers: {
        Location: `https://github.com/${owner}/${repo}/blob/${base}/${year}/post-${now}.md`,
      },
      body: {
        url: `https://github.com/${owner}/${repo}/blob/${base}/${year}/post-${now}.md`,
      },
    };
  } catch (e) {
    return {
      statusCode: 500,
      body: e.toString(),
    };
  }
}

module.exports = { plain, slug };

module.exports.main = wrap(main)
  .with(epsagon)
  .with(statusCheck)
  .with(logger.trace)
  .with(logger);
